### PR TITLE
IP User Component step 2

### PR DESCRIPTION
**Pull request by skorpiom**
Hey,

If you want to test it:

```
npm run start
localhost:3000/users
```
I hardcoded it into that route just for tests, before merging gonna delete (pls don't delete example user).

The importnat part is to set in env:
```
REACT_APP_API_URL=https://coders-camp-organization-app.herokuapp.com
```
ant token with admin role in localStorage

If you want to change user just open MainView component and change it in the function Users.
```
<ManageUser userID="new-id" />
```

Right now that component takes userID as props, but it can  be change into seperate route like a users/:userID if more convienent.

 I added also additional universal button (UButton), just use 'primary' or 'secondary' (warning) color.
```
                              <UButton
                                    text='EDIT'
                                    color='primary'
                                    onClick={toggleEdit}
                                />
```

Tests are on the way.

### CODE REVIEW

**Comment by robert-czwartosz** [comment]
Looks great, but adding grades doesn't work.
... and after changing state of user state, this info disappears.

**Comment by skorpiom** [comment]
@robert-czwartosz thanks, I didn't notice the bug with the user state, most of the development process it worked correctly.

I can't reproduce error with adding grades. 
I suppose that you meant situation, when you want to add grade and the red warning alert appears. 
I improved a little bit that step - you get the orange warning alert 'Section must be selected' and some more appealing UI hints. 

If there is something else I missed please let me know ;)

File: frontend/src/components/FindSection/FindSectionSlice.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  },
+});
+
+export const { } = findSectionSlice.actions;
+
+// if you want, add selectors here, change the one below, remember to register reducer in store.ts
+// export const selectFindSection = (state: RootState) => state.findSection.value;
+
+export default findSectionSlice.reducer;
```
Comment (robert-czwartosz) [inline frontend/src/components/FindSection/FindSectionSlice.ts:25]: I think this file should be removed

File: frontend/src/components/ManageGrades/ManageGradesSlice.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+  },
+});
+
+export const { } = manageGradesSlice.actions;
+
+// if you want, add selectors here, change the one below, remember to register reducer in store.ts
+// export const selectManageGrades = (state: RootState) => state.manageGrades.value;
+
+export default manageGradesSlice.reducer;
```
Comment (robert-czwartosz) [inline frontend/src/components/ManageGrades/ManageGradesSlice.ts:25]: This file is not needed.

File: frontend/src/components/ManageUser/ManageUserSlice.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+  },
+});
+
+export const { } = manageUserSlice.actions;
+
+// if you want, add selectors here, change the one below, remember to register reducer in store.ts
+// export const selectManageUser = (state: RootState) => state.manageUser.value;
+
+export default manageUserSlice.reducer;
```
Comment (robert-czwartosz) [inline frontend/src/components/ManageUser/ManageUserSlice.ts:25]: This slice file looks unneeded

File: frontend/src/components/UButton/UButtonSlice.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  },
+});
+
+export const { } = uButtonSlice.actions;
+
+// if you want, add selectors here, change the one below, remember to register reducer in store.ts
+// export const selectUButton = (state: RootState) => state.uButton.value;
+
+export default uButtonSlice.reducer;
```
Comment (robert-czwartosz) [inline frontend/src/components/UButton/UButtonSlice.ts:25]: This slice file is not needed

File: frontend/src/models/user.model.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+export enum IRole {
+    Candidate,
+    Participant,
+    Mentor,
+    Admin,
+};
+
+export enum IStatus {
```
Comment (robert-czwartosz) [inline frontend/src/models/user.model.ts:8]: What does mean 'I' letter in names IStatus and IRole?
In interfaces 'I' suggests that this is name of interface, but in enum types I dont know what it means.

**Comment by robert-czwartosz** [comment]
File src/components/UButton/UButton.module.css is empty and looks as unneeded.

**Comment by robert-czwartosz** [comment]
@skorpiom If I open select section window, I cannot close it without choosing section.
In addition, there are two warnings in browser console:
```
Warning: findDOMNode is deprecated in StrictMode. findDOMNode was passed an instance of Transition which is inside 
StrictMode. Instead, add a ref directly to the element you want to reference. Learn more about using refs safely here: 
https://reactjs.org/link/strict-mode-find-node
```
and
```
Warning: Each child in a list should have a unique "key" prop.
```

**Review by fszwedo — APPROVED** [review]
codewise nothing I saw to be changed, just one idea - some modal for confirming user deletion might be good, as the action is immediate and user might misslick the delete button or something

File: package-lock.json [hunk 6]
```diff
@@ -1,12669 +0,0 @@
-{
-  "name": "coderscamp2020.project.fullstack-node-react.organizationapp",
-  "version": "1.0.0",
-  "lockfileVersion": 1,
-  "requires": true,
-  "dependencies": {
-    "@babel/code-frame": {
```
Comment (robert-czwartosz) [inline package-lock.json:7]: Why this file was deleted?
Probably this is cause of failing ci check

Comment (skorpiom) [reply package-lock.json:7]: I've got some weird problems with project building process - exactly with Jest (script found version 26.6.3 but react works with 26.6.0). 
I'll add it back soon ;)

**Comment by skorpiom** [comment]
> codewise nothing I saw to be changed, just one idea - some modal for confirming user deletion might be good, as the action is immediate and user might misslick the delete button or something

@fszwedo thanks for suggestion actually I think it's necesary ;)

File: frontend/src/models/user.model.ts (hunk 5 above)
Comment (skorpiom) [reply frontend/src/models/user.model.ts]: You're right, it shouldn't be like that, it introduces confusion and means absolutely nothing - I changed it. Thanks ;)

File: package-lock.json (hunk 6 above)
Comment (robert-czwartosz) [reply package-lock.json:7]: I also had problem with Jest and solved it setting env variable
`SKIP_PREFLIGHT_CHECK=true`

File: frontend/src/components/FindSection/FindSection.tsx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                id: section._id,
+                Name: section.name,
+                "Start date": new Date(section.startDate).toLocaleDateString(),
+                "End date": new Date(section.endDate).toLocaleDateString(),
+            } ));
+        return new Promise(
+            (resolve) => {
+                resolve(sectionsTmp);
+            }
+        );
```
[the remark this answers is not in the document]
Comment (skorpiom) [reply frontend/src/components/FindSection/FindSection.tsx]: right, corrected

File: frontend/src/components/ManageGrades/ManageGrades.tsx [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+    }
+
+    function handleInputChangeGrade(event: React.ChangeEvent<HTMLInputElement>) {
+        const target = event.target;
+        const index = target.tabIndex;
+        const name = target.name;
+        const value = +target.value;
+
+        let tmpGrades: IGrade[] = [ ...grades as IGrade[] ];
+        let grade = tmpGrades[index];
```
[the remark this answers is not in the document]
Comment (skorpiom) [reply frontend/src/components/ManageGrades/ManageGrades.tsx]: right, also changed

File: frontend/src/components/ManageGrades/ManageGrades.tsx [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+    },[]);
+
+    function toggleEdit(index: number) {
+        let edits = [...isEdit];
+        edits[index] = !edits[index];
+        setIsEdit(
+            [...edits],
+        )
+    }
+
```
[the remark this answers is not in the document]
Comment (skorpiom) [reply frontend/src/components/ManageGrades/ManageGrades.tsx:55]: are you sure that it's right? bcs seems like modifying isEdit array (prev state), that breaks the rule of immutability
I was looking for similar examples and always arrays are destructured before modyfing, and here it doesn't seem to be possible, but maybe I miss something ?

File: package-lock.json (hunk 6 above)
Comment (skorpiom) [reply package-lock.json]: right, it works, thanks

File: frontend/src/components/ManageGrades/ManageGrades.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (robert-czwartosz) [reply frontend/src/components/ManageGrades/ManageGrades.tsx:55]: Yeah, my mistake... so I resolve this conversation

File: frontend/src/components/ManageGrades/ManageGrades.tsx [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+    function getGrades(userID: string) {
+        setIsLoaded(false);
+        gradeService.getGrades(userID)
+            .then(res => {
+                if(res.status === 200) {
+                    setIsLoaded(true);
+                    setGrades([...res.data]);
+
+                    let tmpSections: ISectionsUtility[] = res.data.map( (grade: IGrade) => ({_id: grade.sectionId}));
+                    setSections([...tmpSections]);
```
[the remark this answers is not in the document]
Comment (skorpiom) [reply frontend/src/components/ManageGrades/ManageGrades.tsx:99]: I wish I could do like that, but below that expression I invoke function getSectionNames and pass tmpSections argument. 
I know, I could try get sections from state in that function, but it doesn't work in that way - useState hook works kinda async. I could also useEffect hook and pass [sections] dependency but in the getSectionNames function API call is being made so the result would be infinite rerendering, infinite API calls. 

The point is - I need that tmpSections variable.

Async nature of useState also is the problem in getSectionsNames - you suggested to change way of updates - but using that component architecture I don't think it's possible.

**Comment by robert-czwartosz** [comment]
Now component looks and works awesome!
After resolving confilcts feel free to merge

**Review by robert-czwartosz — APPROVED** [review]

**Pull request merged**